mart/attack/perturber.py

```diff
@@ -51,41 +55,103 @@ def __init__(
+        self.projector = projector
+        self.composer = composer
+        self.optim_params = optim_params
+
         self.perturbation = None
 
-    def configure_perturbation(self, input: torch.Tensor | tuple):
-        def create_and_initialize(inp):
-            pert = torch.empty_like(inp, dtype=torch.float, requires_grad=True)
-            self.initializer(pert)
+    def configure_perturbation(self, input: torch.Tensor | tuple | tuple[dict[str, torch.Tensor]]):
+        def create_and_initialize(data, *, input, target, modality="default"):
+            # Though data and target are not used, they are required placeholders for modality_dispatch().
+            pert = torch.empty_like(input, requires_grad=True)
+            self.initializer[modality](pert)
             return pert
 
-        if isinstance(input, tuple):
-            self.perturbation = tuple(create_and_initialize(inp) for inp in input)
-        elif isinstance(input, dict):
-            raise NotImplementedError
+        # Make a dictionary of modality-function.
+        modality_func = {
+            modality: partial(create_and_initialize, modality=modality)
+            for modality in self.initializer
+        }
+
+        # Recursively configure perturbation in tensor.
+        # Though only input=input is used, we have to fill the placeholders of data and target.
+        self.perturbation = modality_dispatch(
+            modality_func, input, input=input, target=input, modality="default"
+        )
+
+    @property
+    def parameter_groups(self):
+        """Extract parameter groups for optimization from perturbation tensor(s)."""
+        param_groups = self._parameter_groups(self.perturbation)
+        return param_groups
+
+    def _parameter_groups(self, pert, modality="default"):
+        """Recursively return parameter groups as a list of dictionaries."""
+
+        if isinstance(pert, torch.Tensor):
+            # Return a list of dictionary instead of a dictionary, easier to extend later.
+            # Add the modality notation so that we can perform gradient modification later.
+            return [{"params": pert, "modality": modality} | self.optim_params[modality]]
+        elif isinstance(pert, dict):
+            param_list = []
+            for modality, pert_i in pert.items():
+                ret_modality = self._parameter_groups(pert_i, modality=modality)
+                param_list.extend(ret_modality)
+            return param_list
+        elif isinstance(pert, (list, tuple)):
+            param_list = []
+            for pert_i in pert:
+                ret_i = self._parameter_groups(pert_i, modality=modality)
+                param_list.extend(ret_i)
+            return param_list
         else:
-            self.perturbation = create_and_initialize(input)
+            raise ValueError(f"Unsupported data type of input: {type(pert)}.")
+
+    def project(self, perturbation, *, input, target, **kwargs):
```

Comment:
Rename to `project_` to adhere to in-place naming scheme.

Reply:
Renamed.